packages/analytics/amplify_analytics_pinpoint_dart/lib/src/analytics_plugin_impl.dart

```diff
@@ -323,4 +321,54 @@ class AmplifyAnalyticsPinpointDart extends AnalyticsPluginInterface {
     await _eventClient.close();
     await _eventStorageAdapter.close();
   }
+
+  /// Retrieve the stored pinpoint endpoint id
+  @visibleForTesting
+  static Future<String> retrieveEndpointId({
+    required String pinpointAppId,
+    required SecureStorageInterface store,
+    LegacyNativeDataProvider? legacyDataProvider,
+  }) async {
+    // Retrieve Unique ID
+    final endpointInformationVersion =
+        await store.read(key: EndpointStoreKey.version.name);
+
+    // Attempt migration if version is null.
+    if (endpointInformationVersion == null) {
+      final legacyEndpointId =
+          await legacyDataProvider?.getEndpointId(pinpointAppId);
+      // Migrate legacy data if it is non-null
+      if (legacyEndpointId != null) {
+        await store.write(
+          key: endpointIdStorageKey,
+          value: legacyEndpointId,
+        );
+      }
+      // Update the version to prevent future legacy data migrations.
+      await store.write(
+        key: EndpointStoreKey.version.name,
+        value: EndpointStoreVersion.v1.name,
+      );
+    }
+
+    // Read the existing ID.
+    final currentEndpointId = await store.read(
+      key: endpointIdStorageKey,
+    );
+
+    final String fixedEndpointId;
+
+    // Generate a new ID if one does not exist.
+    if (currentEndpointId == null) {
+      fixedEndpointId = uuid();
+      await store.write(
+        key: endpointIdStorageKey,
+        value: fixedEndpointId,
+      );
+    } else {
+      fixedEndpointId = currentEndpointId;
+    }
+
+    return fixedEndpointId;
```

Comment:
Since the goal is to arrive at a fixed endpoint ID, use a single variable to convey that journey. This also cleans up the logic branches and saves you a read from the DB in some cases.

```suggestion
    String? fixedEndpointId;
    if (endpointInformationVersion == null) {
      final legacyEndpointId =
          await legacyDataProvider?.getEndpointId(pinpointAppId);
      // Migrate legacy data if it is non-null
      if (legacyEndpointId != null) {
        fixedEndpointId = legacyEndpointId;
        await store.write(
          key: endpointIdStorageKey,
          value: legacyEndpointId,
        );
      }
      // Update the version to prevent future legacy data migrations.
      await store.write(
        key: EndpointStoreKey.version.name,
        value: EndpointStoreVersion.v1.name,
      );
    }

    // Read the existing ID.
    fixedEndpointId ??= await store.read(
      key: endpointIdStorageKey,
    );

    // Generate a new ID if one does not exist.
    if (fixedEndpointId == null) {
      fixedEndpointId = uuid();
      await store.write(
        key: endpointIdStorageKey,
        value: fixedEndpointId,
      );
    }

    return fixedEndpointId;
```

Reply:
My comments are not in conflict with Jordan's. He was confused about the duplication of logic and unnecessary reads/writes.